File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        int prodBefore = 1, prodAfter = 1;
        vector<int> res(n , 1);
        
        for (int i = 1; i < n; i++) {
            prodBefore *= nums[i - 1];
            res[i] = prodBefore;
        }
        
        for (int i = n - 2; i >= 0; i--) {
            prodAfter *= nums[i + 1];
            res[i] *= prodAfter;
        }
        
        return res;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let (mut prodLeft, mut prodRight) = (1, 1);
        let mut res = vec![1; n];
        
        for i in 1..n {
            prodLeft *= nums[i - 1];
            res[i] = prodLeft;
        }
        
        for i in (0..n - 1).rev() {
            prodRight *= nums[i + 1];;
            res[i] *= prodRight;
        }
        res
    }
}